### PR TITLE
add password strength indicator

### LINKED ISSUE #50: Password complexity
http://www.admidio.org/forum/viewtopic.php?f=2&t=6196

I think every password today should have a min of 8 characters (today minlength is 6)
there should be now password maxlength

there are 4 types of characters:
- digits (0-9)
- lower-case letters (a-z)
- upper-case latters (A-Z)
- special chars (-#,.+*=()§$...)

i think every password should contains at least digits and lower or/and upper-case letters

possible solutions:
- option for each category to make it required or not
- change to minlength (but at least 8 chars)
- ~3 different password complexity levels
  - low: 8 chars and contains digits and lower or uppercase letters
  - middle: 10 chars and contains digit and lower and uppercase letters
  - high: 12 chars and contains all 4 types

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/13284455-password-complexity?utm_campaign=plugin&utm_content=tracker%2F10474012&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F10474012&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by ximex**
fixes #50 

TODOs:
- [x] Use const for password-min-length of 8
- [x] Use const for password-gen-length of 16
- [x] Use const for password-gen-chars
- [x] Password-Strength Indicator
- [x] Check password-strength on Server side
- [x] Option to set minimum password-strength
- [ ] Add more user strings to dictionary (optional)
- [ ] More options for password-strength (lower-/uppercase, digits, symbols, minlength) (optional)

### CODE REVIEW

**Comment by ximex** [comment]
@Fasse @PeterTheOne could you look at the code? I think we should make the password strength minimum changeable in the preferences

**Comment by Fasse** [comment]
Yes, it would be nice if the strength parameters could be changed Byte the administrator. Also password lemgth.

I like this feature. Also the indicator.

**Comment by ximex** [comment]
@Fasse the password length in the constants is the really minimum password length and shouldn't be changed. We could set another password length in the preferences that is higher/equal this constant.
Registration is now fixed. little bug.

Missing: option in preferences for password length and maybe more strings for userData.

Possible problem: at registration or if you change your personal data AND your password, the client and server side check could be different. Client could say "strong enough" while Server say "not strong enough".

**Comment by PeterTheOne** [comment]
I haven't really read the code, but it looks good to me.

> Possible problem: at registration or if you change your personal data AND your password, the client and server side check could be different. Client could say "strong enough" while Server say "not strong enough".

I've heard about this problem from another project. Not sure if the problem lies with the data basis or the algorithm (or even the programming language). Setting the client side more strictly might solve it.

**Comment by ximex** [comment]
@Fasse please check and test. if ok you can merge

**Comment by Fasse** [comment]
First test failed: JSON_UNESCAPED_UNICODE is not implemented in 5.3.x

**Comment by ximex** [comment]
@Fasse do you see it? I hate old PHP versions ;-)
My systems are so new that i couldn't easy install a so old Version ;-)

**Comment by Fasse** [comment]
> My systems are so new that i couldn't easy install a so old Version ;-)
Because of this I test your functions ;-)

**Comment by ximex** [comment]
@Fasse could you test if it works if you use `256` instead of `JSON_UNESCAPED_UNICODE`?
`JSON_UNESCAPED_UNICODE` (and `JSON_UNESCAPED_SLASHES`) is available in PHP 5.4+ ...

we have to improve it with this: `JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE`
but first check if it works.

**Comment by Fasse** [comment]
256 works, but did you really test your functionality?

`Notice: Undefined index: user_password in /Users/Jennifer/Sites/github/admidio/adm_program/modules/profile/profile_save.php on line 124`

**Comment by ximex** [comment]
I'm not the great manual tester ;-) hope this fix the last problems

**Comment by Fasse** [comment]
3. Test and I got an error after 1 Test registration!

> Fatal error: Call to a member function format() on a non-object in /Users/Markus/Sites/GitHub/admidio/adm_program/system/classes/user.php on line 1242

Please do at least 1 manual test of your code. Manual tests also belong to software developement and should be done if new features are implemented or bug fixed. It makes little fun to test functions that may never have worked.

**Comment by ximex** [comment]
@Fasse i mostly test only some parts of my code but not every possible combination. but yes i didn't tested the registration.

This should fix it.
```php
    public function getPasswordUserData()
    {
        $userData = array(
            // Names
            $this->getValue('FIRST_NAME'),
            $this->getValue('LAST_NAME'),
            $this->getValue('usr_login_name'),
            // Birthday
            $this->getValue('BIRTHDAY', 'Y'), // YYYY
            $this->getValue('BIRTHDAY', 'md'), // MMDD
            $this->getValue('BIRTHDAY', 'dm'), // DDMM
            // Email
            $this->getValue('EMAIL'),
            // Address
            $this->getValue('ADDRESS'),
            $this->getValue('CITY'),
            $this->getValue('POSTCODE'),
            $this->getValue('COUNTRY')
        );

        /**
         * @param string $value
         * @return bool
         */
        function filterEmptyStrings($value)
        {
            return $value !== '';
        }

        return array_filter($userData, 'filterEmptyStrings');
    }
```

**Comment by ximex** [comment]
i think it should work now. also did some tests ;-)

**Comment by Fasse** [comment]
Registration now works.

First test to change password of login user **admin**:

> Fatal error: Cannot redeclare filterEmptyStrings() (previously declared in /Users/Jennifer/Sites/GitHub/admidio/adm_program/system/classes/user.php:1257) in /Users/Jennifer/Sites/GitHub/admidio/adm_program/system/classes/user.php on line 1257

File: adm_program/languages/de.xml [hunk 1]
```diff
@@ -656,6 +656,13 @@
     <string name="ORG_ORGANIZATION_SUCCESSFULL_ADDED">Die Organisation #VAR1_BOLD# wurde hinzugefügt. Du kannst diese nun bei der Anmeldung auswählen.\n\nFalls du für diese Organisation eine separate Homepage erstellt hast, muss du auf der entsprechenden Seite noch die neue Organisation in die config.php eingetragen.</string>
     <string name="ORG_PARENT_ORGANIZATION">Übergeordnete Organisation</string>
     <string name="ORG_PARENT_ORGANIZATION_DESC">Hier kannst du die übergeordnete Organisation festlegen. Diese haben zum Beispiel die Möglichkeit Termine und Ankündigungen anzulegen, die auch in der aktuellen Organisation sichtbar sind.</string>
+    <string name="ORG_PASSWORD_MIN_STRENGTH">Passwort Sicherheit</string>
+    <string name="ORG_PASSWORD_MIN_STRENGTH_DESC">Hier kannst du festlegen wie sicher/komplex Passwörter mindestens sein müssen.\n\nKeine: riskantes Passwort (&lt;10^3)\nNiedrig: Schutz gegen gedrosselte online Angriffe. (&lt;10^6)\nMittel: Schutz gegen ungedrosselte online Angriffe. (&lt;10^8)\nHoch: Mäßiger schutz gegen offline-hashing Angriffe. (&lt;10^10)\nExtrem: Starker Schutz gegen offline-hashing Angriffe. (&gt;=10^10)</string>
```
Comment (Fasse) [inline adm_program/languages/de.xml]: If we use something like 10^3 or 10^6 than we should explain the meaning of this.

**Comment by Fasse** [comment]
If password is not strong enough you get this message after registration: 

> #PRO_PASSWORDS_NOT_STRONG_ENOUGH#

**Comment by Fasse** [comment]
We should remove the test and demo folders of the new lib **zxcvbn** because this is not necessary and could contain pages that users should not view within our installations even if demo pages contain security vulnerabilities.

We should also add a simple **admidio_update.txt** like in **datatables** to explain the steps to update this lib.

File: adm_program/languages/de.xml (hunk 1 above)
Comment (ximex) [inline adm_program/languages/de.xml]: `In () die geschätzten Versuche bis geknackt.` as explanation?

**Comment by Fasse** [comment]
- Why not add the strength indicator during installation for admin?
- Within string ORG_PASSWORD_MIN_STRENGTH_DESC you use 10^x ... please explain the meaning in the string or remove it, because not everyone know this.
- Last but not least we should add the lib to our readme.md.

**Comment by ximex** [comment]
> - Within string ORG_PASSWORD_MIN_STRENGTH_DESC you use 10^x ... please explain the meaning in the string or remove it, because not everyone know this.

look at: https://github.com/Admidio/admidio/pull/378/commits/c7fd3a6b46f619a145e6bcd74b392f6033067808#diff-2f5f43cf2b8e5fd1a4d4b54ca788e665R660

other things should be done in last commit

**Comment by PeterTheOne** [comment]
Yay, merge. :tada:

**Pull request merged**